pkg/apis/resource/validation/validation.go

```diff
@@ -200,6 +202,12 @@ func validateResourceHandles(resourceHandles []resource.ResourceHandle, maxSize
 		if len(resourceHandle.Data) > resource.ResourceHandleDataMaxSize {
 			allErrs = append(allErrs, field.TooLongMaxLength(idxPath.Child("data"), len(resourceHandle.Data), resource.ResourceHandleDataMaxSize))
 		}
+		if resourceHandle.StructuredData != nil {
+			allErrs = append(allErrs, validateStructuredResourceHandle(resourceHandle.StructuredData, idxPath.Child("structuredData"))...)
+		}
+		if len(resourceHandle.Data) > 0 && resourceHandle.StructuredData != nil {
+			allErrs = append(allErrs, field.Invalid(idxPath, nil, "both data and structuredData are set although they are mutually exclusive"))
```

Comment:
It's a nit, but usually we would say

```
"must not be specified when `data` is set"
```

or

```
"data and structuredData are mutually exclusive"
```

Reply:
The latter. Changed